File: hs/src/Day25.hs
{-|
Module:         Day25
Description:    <https://adventofcode.com/2022/day/25 Day 25: Full of Hot Air>
-}
module Day25 (day25) where

import Data.Char (digitToInt, intToDigit)
import Data.Text (Text)
import qualified Data.Text as T (foldl', lines, unfoldr, reverse)

day25 :: Text -> Text
day25 = T.reverse . T.unfoldr g . sum . map (T.foldl' f 0) . T.lines where
    f k '=' = 5 * k - 2
    f k '-' = 5 * k - 1
    f k c = 5 * k + digitToInt c
    g 0 = Nothing
    g n = Just $ case n `divMod` 5 of
        (q, 3) -> ('=', q + 1)
        (q, 4) -> ('-', q + 1)
        (q, r) -> (intToDigit r, q)

File: py/aoc2022/day25.py
"""
Day 25: Full of Hot Air
"""

SAMPLE_INPUT = [
    "1=-0-2",
    "12111",
    "2=0=",
    "21",
    "2=01",
    "111",
    "20012",
    "112",
    "1=-1=",
    "1-12",
    "12",
    "1=",
    "122",
]


def _unsnafu(snafu):
    number = 0
    for char in snafu:
        number = 5 * number + (-2 if char == "=" else -1 if char == "-" else int(char))
    return number


def _snafu(number):
    snafu = ""
    while number:
        snafu += "012=-"[number % 5]
        number = (number + 2) // 5
    return snafu[::-1]


def part1(lines):
    """
    >>> part1(SAMPLE_INPUT)
    '2=-1=0'
    """
    return _snafu(sum(_unsnafu(line.rstrip()) for line in lines))


parts = (part1,)
